stlab/channel.hpp

```diff
@@ -583,12 +631,211 @@ struct shared_process : shared_process_receiver<yield_type<T, Arg>>,
+auto create_receiver_collector(std::shared_ptr<C>& context, R&& upstream_receiver) {
+    using context_worker = typename W::template type<I, C>;
+    auto receiver_collector = std::forward<R>(upstream_receiver) | context_worker(context);
+    return receiver_collector._p;
+}
+
+template <typename W, typename C, typename... R, std::size_t... I>
+auto create_receiver_collectors(std::shared_ptr<C>& context, std::index_sequence<I...>, R&&... upstream_receiver) {
+    return std::make_tuple(create_receiver_collector<I,W>(context, std::forward<R>(upstream_receiver))...);
+}
+
+template <typename C, typename S, typename P, typename RC, std::size_t... I>
+auto create_process(S&& s, P&& p, RC& receiver_collectors, std::index_sequence<I...>) {
+    return std::make_shared<detail::shared_process<P, C>>(
+        std::forward<S>(s), std::forward<P>(p), std::get<I>(receiver_collectors)...);
+}
+
+template <typename C, typename RC, typename P, std::size_t... I>
+void map_as_sender_(RC& receiver_collectors, P& p, std::index_sequence<I...>) {
+    (void)std::initializer_list<int>{(std::get<I>(receiver_collectors)->map(sender<C>(p)), 0)...};
+}
+
+template <typename C, typename RC, typename P>
+void map_as_sender(RC& receiver_collectors, P& p) {
+    map_as_sender_<C>(receiver_collectors, p, std::make_index_sequence<std::tuple_size<RC>::value>());
+}
+
+template <typename F, typename...T, std::size_t... I>
+auto apply_join_args_(const F& f, std::tuple<boost::optional<T>...>& args, std::index_sequence<I...>) {
+    return f(std::move(std::get<I>(args)).value()...);
+}
+
+template <typename F, typename... T>
+auto apply_join_args(const F& f, std::tuple<boost::optional<T>...>& args) {
+    return apply_join_args_(f, args, std::make_index_sequence<sizeof...(T)>());
+}
+
+template <std::size_t I, typename... T>
+void reset_tuple_optional_(std::tuple<boost::optional<T>...>& values) {
+    std::get<I>(values) = boost::none;
+}
+
+template <typename... T, std::size_t... I>
+void reset_tuple_optional(std::tuple<boost::optional<T>...>& values, std::index_sequence<I...>) {
+    (void)std::initializer_list<int>{(reset_tuple_optional_<I>(values), 0)...};
+}
+
+
+
+template <typename... T>
+struct join_context
+{
+    const static std::size_t            _size = sizeof...(T);
+    std::tuple<boost::optional<T>...>   _result;
+    std::atomic_size_t                  _remaining{ _size };
+
+    template<std::size_t i, typename U>
+    void await(U&& u) {
+        assert(!std::get<i>(_result).is_initialized());
+        std::get<i>(_result) = std::forward<U>(u);
+    }
+
+    void reset() {
+        reset_tuple_optional(_result, std::make_index_sequence<sizeof...(T)>());
+        _remaining = _size;
+    }
+
+    std::tuple<boost::optional<T>...> yield() {
+        if (--_remaining == 0) {
+            auto result = std::move(_result);
+            reset();
+            return result;
+        }
+        return _result;
```

Comment:
This seems very suspect. If we haven't received all values, then the process needs to appear to be still running to any senders that have not sent a value (their cts() has not been called) but to any senders that haven't yet sent a value we are ready to receive. Once we have all awaited values, then we change to a yield state. I think to make this work more of the logic goes to the shared_process.